demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c

```diff
@@ -0,0 +1,1608 @@
+
+static int tlsSetup( int tcpSocket,
+                     const char * pAlpnProtos,
+                     size_t alpnProtosLen,
+                     SSL ** pSslContext )
+{
+    int status = EXIT_FAILURE, sslStatus = 0, alpnStatus = -1;
+    FILE * pRootCaFile = NULL;
+    X509 * pRootCa = NULL;
+    SSL_CTX * pSslSetup = NULL;
+
+    assert( tcpSocket >= 0 );
+
+    /* Setup for creating a TLS client. */
+    pSslSetup = SSL_CTX_new( TLS_client_method() );
+
+    if( pSslSetup != NULL )
+    {
+        /* Set auto retry mode for the blocking calls to SSL_read and SSL_write.
+         * The mask returned by SSL_CTX_set_mode does not need to be checked. */
+        ( void ) SSL_CTX_set_mode( pSslSetup, SSL_MODE_AUTO_RETRY );
+
+        /* Setup authentication. */
+        sslStatus = readCredentials( pSslSetup,
+                                     ROOT_CA_CERT_PATH,
+                                     CLIENT_CERT_PATH,
+                                     CLIENT_PRIVATE_KEY_PATH );
+
+        if( sslStatus != 1 )
+        {
+            LogError( ( "Setting up credentials failed." ) );
+        }
+    }
+
+    /* Set up the TLS connection. */
+    if( sslStatus == 1 )
+    {
+        /* Create a new SSL context. */
+        *pSslContext = SSL_new( pSslSetup );
+
+        if( *pSslContext != NULL )
+        {
+            /* Enable SSL peer verification. */
+            SSL_set_verify( *pSslContext, SSL_VERIFY_PEER, NULL );
+            sslStatus = SSL_set_fd( *pSslContext, tcpSocket );
+
+            if( sslStatus != 1 )
+            {
+                LogError( ( "Failed to set the socket fd to SSL context." ) );
+            }
+        }
+        else
+        {
+            LogError( ( "Failed to create a new SSL context." ) );
+            sslStatus = 0;
+        }
+    }
+
+    if( ( sslStatus == 1 ) &&
+        ( pAlpnProtos != NULL ) && ( alpnProtosLen > 0 ) )
+    {
+        sslStatus = -1;
+
+        if( AWS_MQTT_PORT != 443 )
+        {
+            LogError( ( "AWS_MQTT_PORT must be set to 443 to use ALPN." ) );
+        }
+        else
+        {
+            alpnStatus = SSL_set_alpn_protos( *pSslContext,
+                                              ( unsigned char * ) pAlpnProtos,
+                                              ( unsigned int ) alpnProtosLen );
+        }
+
+        if( alpnStatus != 0 )
+        {
+            LogError( ( "SSL_set_alpn_protos failed to set ALPN protos. %s",
+                        pAlpnProtos ) );
+        }
+        else
+        {
+            sslStatus = 1;
+        }
+    }
+
+    /* Perform the TLS handshake. */
+    if( sslStatus == 1 )
+    {
+        sslStatus = SSL_connect( *pSslContext );
+
+        if( sslStatus != 1 )
+        {
+            LogError( ( "Failed to perform TLS handshake." ) );
+        }
+    }
+
+    /* Verify the peer certificate. */
+    if( sslStatus == 1 )
+    {
+        if( SSL_get_verify_result( *pSslContext ) != X509_V_OK )
```

Comment:
Same here - store the function call result in a local variable and compare that in the if condition. This is helpful while debugging.